### PR TITLE
LXD storage api: Implement storage volume renaming

### LINKED ISSUE #2865: LXD storage api: Implement storage volume renaming if applicable
Title says it all.

### PR DESCRIPTION
**Pull request by brauner**
Closes #2865.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by stgraber** [comment]
This needs rebasing.

File: lxd/storage_volumes.go [hunk 1]
```diff
@@ -123,7 +126,9 @@ func storagePoolVolumesTypeGet(d *Daemon, r *http.Request) Response {
 
 			volumeUsedBy, err := storagePoolVolumeUsedByGet(d.State(), vol.Name, vol.Type)
 			if err != nil {
-				return SmartError(err)
+				if !os.IsNotExist(err) {
```
Comment (stgraber) [inline lxd/storage_volumes.go]: Why the nested if statement?

Comment (stgraber) [reply lxd/storage_volumes.go]: Same comment applies to a number of similar checks above.

File: lxd/storage_volumes.go [hunk 2]
```diff
@@ -187,6 +192,66 @@ func storagePoolVolumesTypePost(d *Daemon, r *http.Request) Response {
...
+	if req.Name == "" {
+		return BadRequest(fmt.Errorf("No name provided"))
+	}
+
+	// We currently only allow to create storage volumes of type
+	// storagePoolVolumeTypeCustom. So check, that nothing else was
+	// requested.
+	if volumeTypeName != storagePoolVolumeTypeNameCustom {
+		return BadRequest(fmt.Errorf("Currently not allowed to create storage volumes of type %s", volumeTypeName))
+	}
```
Comment (stgraber) [inline lxd/storage_volumes.go]: The error string seems wrong.

File: lxd/storage_volumes_utils.go [hunk 3]
```diff
@@ -200,6 +201,10 @@ func storagePoolVolumeUsedByContainersGet(s *state.State, volumeName string,
 		}
 	}
 
+	if len(ctsUsingVolume) == 0 {
+		return []string{}, os.ErrNotExist
```
Comment (stgraber) [inline lxd/storage_volumes_utils.go]: That seems wrong to me, it looks to me like the pool doesn't exist when in fact it's just that we don't have any volume, why isn't an empty list the right answer here?

File: lxd/storage_volumes_utils.go [hunk 4]
```diff
@@ -240,7 +245,7 @@ func storagePoolVolumeUsedByGet(s *state.State, volumeName string, volumeTypeNam
 	}
 
 	if len(volumeUsedBy) == 0 && len(profiles) == 0 {
-		return []string{}, err
+		return []string{}, os.ErrNotExist
```
Comment (stgraber) [inline lxd/storage_volumes_utils.go]: Same as above

File: client/interfaces.go [hunk 5]
```diff
@@ -152,6 +152,7 @@ type ContainerServer interface {
 	CreateStoragePoolVolume(pool string, volume api.StorageVolumesPost) (err error)
 	UpdateStoragePoolVolume(pool string, volType string, name string, volume api.StorageVolumePut, ETag string) (err error)
 	DeleteStoragePoolVolume(pool string, volType string, name string) (err error)
+	RenameStoragePoolVolume(pool string, volType string, name string, volume api.StorageVolumesPost) (err error)
```
Comment (stgraber) [inline client/interfaces.go]: This is wrong, the argument should be a new api.StorageVolumePost struct.

File: lxd/storage_volumes.go (hunk 2 above)
Comment (brauner) [reply lxd/storage_volumes.go]: Yep

File: lxd/storage_volumes_utils.go (hunk 3 above)
Comment (brauner) [reply lxd/storage_volumes_utils.go]: I'd prefer notexist errors but I'm fine the other way too.

File: lxd/storage_volumes_utils.go (hunk 4 above)
Comment (brauner) [reply lxd/storage_volumes_utils.go]: I don't like this checking for empty entities kind of stuff but as I said I'm fine with both.

**Comment by stgraber** [comment]
I think I prefer functions like storagePoolVolumeUsedByGet to return an empty list if nothing is using it. Errors should be kept for errors. A NotExist error to me would only make sense if the pool or volume didn't exist.

File: shared/api/storage_pool_volume.go [hunk 6]
```diff
@@ -10,6 +10,16 @@ type StorageVolumesPost struct {
 	Type string `json:"type" yaml:"type"`
 }
 
+// StorageVolumePost represents the fields required to rename a LXD storage pool volume
+//
+// API extension: storage_api_volume_rename
+type StorageVolumePost struct {
+	StorageVolumePut `yaml:",inline"`
```
Comment (stgraber) [inline shared/api/storage_pool_volume.go]: This is wrong, the rename struct should only contain the new Name (and once we add support for moving pools, the new pool).

File: shared/api/storage_pool_volume.go [hunk 7]
```diff
@@ -10,6 +10,16 @@ type StorageVolumesPost struct {
...
+
+	Name string `json:"name" yaml:"name"`
+	Type string `json:"type" yaml:"type"`
```
Comment (stgraber) [inline shared/api/storage_pool_volume.go]: Same comment as above, shouldn't be in StorageVolumePost.

**Comment by stgraber** [comment]
Just two small cleanups to do in shared/api and this will be good to go.

**Comment by stgraber** [comment]
This isn't building anymore :)

**Comment by brauner** [comment]
@stgraber, should be good to go.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
I was about to hit merge and then I noticed that this is missing a doc/rest-api.md update for the new API endpoint...

**Comment by brauner** [comment]
@stgraber, added.

**Pull request merged**